pkg/disk/waitstatus/batched.go

```diff
@@ -0,0 +1,216 @@
+		ecsClient:   ecsClient,
+		requestChan: make(chan *waitRequest[*T]),
+		requests:    make(map[string][]*waitRequest[*T]),
+		feedback:    make(chan responseFeedback[T]),
+
+		defaultInterval: defaultInterval,
+		maxInterval:     maxInterval,
+		clk:             clk,
+	}
+}
+
+func (w *Batched[T]) WaitFor(ctx context.Context, id string, pred StatusPredicate[*T]) (*T, error) {
+	resultChan := make(chan waitResponse[*T], 1) // in case this wait is canceled, don't block poll()
+	w.requestChan <- &waitRequest[*T]{
+		ctx:        ctx,
+		id:         id,
+		pred:       pred,
+		resultChan: resultChan,
+	}
+	select {
+	case resp := <-resultChan:
+		return resp.res, resp.err
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	}
+}
+
+func (w *Batched[T]) enqueueRequest(r *waitRequest[*T]) {
+	r.time = w.clk.Now()
+	reqs, ok := w.requests[r.id]
+	w.requests[r.id] = append(reqs, r)
+	if !ok {
+		w.idQueue = append(w.idQueue, r.id)
+	}
+}
+
+func (w *Batched[T]) Run(ctx context.Context) {
+	var pollChan <-chan time.Time
+	var lastPollTime time.Time
+	logger := klog.FromContext(ctx).WithValues("type", w.ecsClient.Type())
+
+	for {
+		if pollChan == nil && len(w.idQueue) > 0 {
+			// If this is the first request, poll immediately.
+			nextPoll := lastPollTime.Add(w.defaultInterval)
+			upperBound := w.requests[w.idQueue[0]][0].time.Add(w.maxInterval)
+			if upperBound.Before(nextPoll) {
+				nextPoll = upperBound
+			}
+			d := -w.clk.Since(nextPoll)
+			pollChan = w.clk.After(d)
+			logger.V(3).Info("poll scheduled", "after", d, "queueDepth", len(w.idQueue))
+		}
+
+		select {
+		case <-ctx.Done():
+			return
+		case r := <-w.requestChan:
+			klog.FromContext(r.ctx).V(5).Info("enqueued waitstatus", "type", w.ecsClient.Type())
+			w.enqueueRequest(r)
+		case r := <-w.feedback:
+			next := w.processFeedback(r)
+			w.idQueue = append(w.idQueue, next...)
+			logger.V(4).Info("poll response processed", "queueDepth", len(w.idQueue), "requeue", len(next))
+		case t := <-pollChan:
+			logger.V(4).Info("starting poll", "queueDepth", len(w.idQueue))
+			w.idQueue = w.poll(t, w.idQueue)
+			lastPollTime = t
+			pollChan = nil
+		}
+	}
+}
+
+func (w *Batched[T]) got(t time.Time, id string, resource *T, requestID string) (done bool) {
+	pendingReqs := slices.DeleteFunc(w.requests[id], func(r *waitRequest[*T]) bool {
+		if r.ctx.Err() != nil {
+			return true
+		}
+		logger := klog.FromContext(r.ctx)
+		if resource == nil || r.pred(resource) {
+			logger.V(4).Info("reached desired status", "type", w.ecsClient.Type(), "requestID", requestID)
+			r.resultChan <- waitResponse[*T]{res: resource}
+			return true
+		}
+		logger.V(4).Info("status not reached, re-queue", "type", w.ecsClient.Type(), "requestID", requestID)
+		r.time = t
+		return false
+	})
+	if len(pendingReqs) == 0 {
+		delete(w.requests, id)
+		return true
+	} else {
+		w.requests[id] = pendingReqs
+		return false
+	}
+}
+
+func (w *Batched[T]) processFeedback(fb responseFeedback[T]) (next []string) {
```

Comment:
Why not eliminate the feedback channel and incorporate the response directly into resultChan?

Reply:
Because I want to avoid a mutex to protect `w.requests` and `w.idQueue`. I found it easier to  understand if I only modify `w` from one goroutine.

Reply:
okay, gotcha, please add the explanation